### PR TITLE
driver(docker): opt to set additional dial meta to the client

### PR DESCRIPTION
**Pull request by crazy-max**
if you want some changes apart, it then needs:
* closes https://github.com/docker/buildx/pull/2074
* closes https://github.com/docker/buildx/pull/2075

In some cases it can be useful to set additional dial metadata to the client.

### CODE REVIEW

File: driver/docker-container/driver.go [hunk 1]
```diff
@@ -380,9 +387,19 @@ func (d *Driver) Client(ctx context.Context) (*client.Client, error) {
...
 	}))
+
 	if td, ok := exp.(client.TracerDelegate); ok {
 		opts = append(opts, client.WithTracerDelegate(td))
 	}
+
+	if len(co.Meta) > 0 {
+		opts = append(opts, client.WithGRPCDialOption(grpc.WithChainUnaryInterceptor(func(ctx context.Context, method string, req, reply interface{}, cc *grpc.ClientConn, invoker grpc.UnaryInvoker, opts ...grpc.CallOption) error {
+			// merge the existing context with new metadata.
+			ctx = metadata.NewOutgoingContext(ctx, co.Meta)
```
Comment (crazy-max) [inline driver/docker-container/driver.go:398]: I think it's safe to make a new set of metadata but let me know if you prefer to append instead for extra safety.

File: driver/kubernetes/driver.go [hunk 2]
```diff
@@ -218,9 +225,19 @@ func (d *Driver) Client(ctx context.Context) (*client.Client, error) {
 	opts = append(opts, client.WithContextDialer(func(context.Context, string) (net.Conn, error) {
 		return conn, nil
 	}))
+
 	if td, ok := exp.(client.TracerDelegate); ok {
 		opts = append(opts, client.WithTracerDelegate(td))
 	}
+
+	if len(co.Meta) > 0 {
+		opts = append(opts, client.WithGRPCDialOption(grpc.WithChainUnaryInterceptor(func(ctx context.Context, method string, req, reply interface{}, cc *grpc.ClientConn, invoker grpc.UnaryInvoker, opts ...grpc.CallOption) error {
```
Comment (jedevc) [inline driver/kubernetes/driver.go:234]: I don't have all the context for this, but I think we probably want to add this to `grpc.WithChainStreamInterceptor` as well?

File: driver/docker-container/driver.go (hunk 1 above)
Comment (crazy-max) [reply driver/docker-container/driver.go:398]: Actually it's not safe, we need to append as we are joining metadata in buildkit like `buildkit-attachable-store-id`. Opened https://github.com/moby/buildkit/pull/4175.

File: driver/kubernetes/driver.go (hunk 2 above)
Comment (crazy-max) [reply driver/kubernetes/driver.go]: As discussed with @tonistiigi, drop support for other drivers for now.

File: driver/driver.go [hunk 3]
```diff
@@ -110,3 +110,15 @@ func historyAPISupported(ctx context.Context, c *client.Client) bool {
 		}
 	}
 }
+
+type ClientOption func(*ClientOptions)
+
+type ClientOptions struct {
+	Meta map[string][]string
```
Comment (tonistiigi) [inline driver/driver.go:117]: I think `Meta` is too generic in here. `TransportMeta` or `DialMeta`.

**Review by tonistiigi — COMMENTED** [review]
Bring this out of draft if ready.

**Review by tonistiigi — APPROVED** [review]

**Pull request merged**